### PR TITLE
Extend Virtual Inputs to allow digests

### PR DESCRIPTION
**Pull request by andusy**
Allows virtual inputs to be used with digests. If Contents is provided with InputDigest or the digest does not exist in the CAS, return an error.

Bug: b/324349357, https://github.com/bazelbuild/remote-apis-sdks/issues/525

### CODE REVIEW

File: go/pkg/client/tree.go [hunk 1]
```diff
@@ -440,9 +440,21 @@ func (c *Client) ComputeMerkleTree(ctx context.Context, execRoot, workingDir, re
 			}
 			continue
 		}
+		if i.InputDigest != "" && i.Contents != nil {
+			return digest.Empty, nil, nil, errors.New("digest and file content cannot be provided for the same virtual input")
+		}
+		entry := uploadinfo.EntryFromBlob(i.Contents)
```
Comment (mrahs) [inline go/pkg/client/tree.go:446]: It's worth placing this in an `else` branch to make it a bit clearer that either bytes or hash is used. It also avoids the cost of creating unnecessary structs.

File: go/pkg/client/tree.go [hunk 2]
```diff
@@ -440,9 +440,21 @@ func (c *Client) ComputeMerkleTree(ctx context.Context, execRoot, workingDir, re
...
+		if i.InputDigest != "" {
+			dg, err := digest.NewFromString(i.InputDigest)
+			if err != nil {
+				return digest.Empty, nil, nil, err
+			}
+			entry = uploadinfo.EntryFromFile(dg, normPath)
```
Comment (mrahs) [inline go/pkg/client/tree.go:452]: Why mark it as a file blob rather than a bytes blob?

File: go/pkg/client/cas_upload.go [hunk 3]
```diff
@@ -287,6 +287,9 @@ func (c *Client) uploadUnified(ctx context.Context, entries ...*uploadinfo.Entry
 			contextmd.Infof(ctx, log.Level(2), "Skipping upload of empty entry %s", ue.Digest)
 			continue
 		}
+		if ue.IsVirtualInput && ue.IsFile() {
+			return nil, 0, fmt.Errorf("virtual input with digest: %s provided, but does not exist in CAS", ue.Digest)
+		}
```
Comment (mrahs) [inline go/pkg/client/cas_upload.go:292]: This block is not clear to me. Why is it an error to have a missing virtual input?

File: go/pkg/client/tree.go (hunk 1 above)
Comment (andusy) [reply go/pkg/client/tree.go:446]: I wanted to preserve the current behaviour. We don't actually check that the Contents are not empty before creating an entry. It looks like we just check that the digest is empty before downloading https://github.com/bazelbuild/remote-apis-sdks/blob/master/go/pkg/client/cas_upload.go#L286

File: go/pkg/client/cas_upload.go (hunk 3 above)
Comment (andusy) [reply go/pkg/client/cas_upload.go:292]: VirtualInputs should exist on RBE and if it doesn't then it means something likely already went wrong before reaching this step. I don't think we should be uploading a possibly corrupted digest in this case.

File: go/pkg/client/tree.go (hunk 2 above)
Comment (andusy) [reply go/pkg/client/tree.go:452]: I'm not too familiar with specifics within the sdk. Is there a behavioural difference when marking an input as file vs a blob? Which would you suggest to use in this case?

Comment (mrahs) [reply go/pkg/client/tree.go:452]: The path field should point to the absolute local path. Virtual inputs do not exist locally by definition so they should all be of type blob.

File: go/pkg/client/tree.go (hunk 1 above)
Comment (mrahs) [reply go/pkg/client/tree.go:446]: The `if-else` block would only alter the source of the entry. If the digest is not specified, the entry will still be created from the content regardless of its actual value.

File: go/pkg/client/tree.go (hunk 2 above)
Comment (mrahs) [reply go/pkg/client/tree.go:452]: Hmm, but now that we are changing that definition to include two more cases:
- files that exist locally but the digest for which is already computed
- files that do not exist locally but we know their digest and it should already exist remotely

I think the implementation should take all three cases into account.

Comment (andusy) [reply go/pkg/client/tree.go]: Going to label the cases you provided to make it easier:
1. files that exist locally but the digest for which is already computed
2. files that do not exist locally but we know their digest and it should already exist remotely

(1) In this case VirtualInputs take precedence, correct? Meaning, since we ignore the local file this would technically just be treated as (2)?

Comment (mrahs) [reply go/pkg/client/tree.go]: I agree. It's simpler to define just two disjoint cases:
1. bytes (no digest) that can be digested and uploaded if necessary
2. a digest (no bytes) that should exist remotely

File: go/pkg/client/tree.go [hunk 4]
```diff
@@ -440,9 +440,23 @@ func (c *Client) ComputeMerkleTree(ctx context.Context, execRoot, workingDir, re
...
+		if i.InputDigest != "" && len(i.Contents) > 0 {
+			return digest.Empty, nil, nil, errors.New("digest and file content cannot be provided for the same virtual input")
+		}
+		var entry *uploadinfo.Entry
+		if i.InputDigest != "" {
+			dg, err := digest.NewFromString(i.InputDigest)
+			if err != nil {
+				return digest.Empty, nil, nil, err
+			}
+			entry = uploadinfo.EntryFromRemoteFile(dg, normPath)
```
[the remark this answers is not in the document]
Comment (mrahs) [reply go/pkg/client/tree.go]: You're right. I read that wrong. It's usually the absolute path if I'm not mistaken.

File: go/pkg/client/cas_upload.go [hunk 5]
```diff
@@ -287,6 +287,9 @@ func (c *Client) uploadUnified(ctx context.Context, entries ...*uploadinfo.Entry
 			contextmd.Infof(ctx, log.Level(2), "Skipping upload of empty entry %s", ue.Digest)
 			continue
 		}
+		if ue.IsVirtualFile() {
+			return nil, 0, fmt.Errorf("virtual input with digest: %s provided, but does not exist in CAS", ue.Digest)
```
Comment (mrahs) [inline go/pkg/client/cas_upload.go:291]: nit: 
```suggestion
			return nil, 0, fmt.Errorf("virtual input with digest %q provided, but does not exist in CAS", ue.Digest)
```

File: go/pkg/uploadinfo/entry.go [hunk 6]
```diff
@@ -59,3 +65,14 @@ func EntryFromFile(dg digest.Digest, path string) *Entry {
...
 	}
 }
+
+// EntryFromVirtualFile creates an entry from a file not on disk.
+// The digest is expected to exist in the CAS.
+func EntryFromVirtualFile(dg digest.Digest, path string) *Entry {
+	return &Entry{
+		Digest:      dg,
+		Path:        path,
+		ueType:      ueBlob,
```
Comment (mrahs) [inline go/pkg/uploadinfo/entry.go:75]:
```suggestion
		ueType:      ueFile,
```

File: go/pkg/client/tree_test.go [hunk 7]
```diff
@@ -1565,6 +1590,10 @@ func TestComputeMerkleTree(t *testing.T) {
 				t.Errorf("ComputeMerkleTree(...) = gave error %q, want success", err)
 			}
 			for _, ue := range inputs {
+				if ue.IsVirtualFile() {
+					gotBlobs[ue.Digest] = tc.digesttoBlob[ue.Digest]
```
Comment (mrahs) [inline go/pkg/client/tree_test.go:1594]: Why is this needed?

File: go/pkg/uploadinfo/entry.go (hunk 6 above)
Comment (andusy) [reply go/pkg/uploadinfo/entry.go:75]: Don't we want to use blobs for virtual files?

File: go/pkg/client/tree_test.go (hunk 7 above)
Comment (andusy) [reply go/pkg/client/tree_test.go:1594]: chunker.New tries to:
 - read the content for blobs
 - read the file for files
 
Both of which won't be available in this case. So for testing purposes I'm just mapping the virtual inputs' digest back to the corresponding blob.

Comment (mrahs) [reply go/pkg/client/tree_test.go:1594]: Would it work if you excluded virtual lobs from `additionalBlobs` instead?

File: go/pkg/uploadinfo/entry.go (hunk 6 above)
Comment (mrahs) [reply go/pkg/uploadinfo/entry.go:75]: My suggestion was from a consistency perspective. Since those are not meant to be uploaded, it doesn't matter what the type is.

File: go/pkg/client/tree_test.go (hunk 7 above)
Comment (andusy) [reply go/pkg/client/tree_test.go]: yes, but I would still have to add a condition here since `ueFile` would not be (and probably should not be) handled in `chunker.New()`.

Comment (mrahs) [reply go/pkg/client/tree_test.go]: That's what I had in mind. Thanks.

File: go/pkg/client/tree_test.go [hunk 8]
```diff
@@ -462,6 +462,7 @@ func TestComputeMerkleTree(t *testing.T) {
 		// blobs.
 		rootDir         *repb.Directory
 		additionalBlobs [][]byte
+		digesttoBlob    map[digest.Digest][]byte
```
Comment (mrahs) [inline go/pkg/client/tree_test.go:465]: No longer needed.

File: go/pkg/client/tree_test.go [hunk 9]
```diff
@@ -1462,6 +1463,30 @@ func TestComputeMerkleTree(t *testing.T) {
...
+				InputNodeProperties: map[string]*cpb.NodeProperties{"fooDir/foo": fooProperties},
+			},
+			rootDir: &repb.Directory{Directories: []*repb.DirectoryNode{
+				{Name: "barDir", Digest: barDirDgPb},
+				{Name: "fooDir", Digest: fooDirDgPb},
+			}},
+			additionalBlobs: [][]byte{fooDirBlob, barDirBlob},
+			digesttoBlob: map[digest.Digest][]byte{
+				fooDg: fooBlob,
+				barDg: barBlob,
```
Comment (mrahs) [inline go/pkg/client/tree_test.go:1482]: leftover

File: go/pkg/uploadinfo/entry.go [hunk 10]
```diff
@@ -10,6 +10,7 @@ import (
 const (
 	ueBlob = iota
 	uePath
+	ueFile
```
Comment (mrahs) [inline go/pkg/uploadinfo/entry.go:13]: Ah, my bad. I meant to type `uePath`. No need for a new type.

**Review by mrahs — APPROVED** [review]
